### PR TITLE
Read items within a collection from kopia on demand

### PR DESCRIPTION
**Pull request by meain**
Don't read all the files that we need to restore, just the collections in the beginning. The rest of them are read using Fetch on demand when they will be restored.

I haven't tested for any changes in memory consumption, but this brings down time taken for "Enumerating items in repository" to <5s even for huge number of files.

---

#### Does this PR need a docs update or release note?

- [ ] :white_check_mark: Yes, it's included
- [x] :clock1: Yes, but in a later PR
- [ ] :no_entry: No

#### Type of change

- [x] :sunflower: Feature
- [ ] :bug: Bugfix
- [ ] :world_map: Documentation
- [ ] :robot: Supportability/Tests
- [ ] :computer: CI/Deployment
- [ ] :broom: Tech Debt/Cleanup

#### Issue(s)

* https://github.com/alcionai/corso/issues/3011
* closes https://github.com/alcionai/corso/issues/3440
* closes https://github.com/alcionai/corso/issues/3537

#### Test Plan

- [x] :muscle: Manual
- [ ] :zap: Unit test
- [x] :green_heart: E2E

### CODE REVIEW

**Review by ashmrtn — COMMENTED** [review]
generally looks good. Just needs some fixups for the `Items` unit test

File: src/internal/kopia/data_collection.go
Comment (vkamra) [reply src/internal/kopia/data_collection.go:58]: 👍 - Yep - @meain had "volunteered" to pick this up to help with the memory issue.

Comment (vkamra) [reply src/internal/kopia/data_collection.go:58]: I can do some profiling also later today to check on this.

**Comment by ashmrtn** [comment]
ah, just remembered this, but since this PR changes the semantics of loading files (we get not found errors during `Items` now) we need to double check it's not going to adversely affect incremental backups. The two areas to focus on would be (numbered for reference, not importance)
1. how we select the base. We'll no longer get errors here which should be fine but we'll want to double check prior to merging
2. how we load the previous paths and delta tokens. Now that we may get notifications that stuff wasn't found later we need to be sure we can handle partial results during deserialization. I think we can but should double check
3. ensure that having some errors in the fault bus doesn't muck up later parts of the backup. Not really sure how we should handle CLI display on this part though. We may need to consider making local fault busses when populating the previous paths and delta token maps

Since I'm reworking the logic for selecting base backups it may be best to put (1) off until we're closer to merging this PR

**Review by ashmrtn — COMMENTED** [review]
PR looks ready save for one failing non-unit test in wrapper_test.go

going to hold off on approval until we have more clarity on how this will impact other parts of the system

File: src/internal/kopia/data_collection.go
Comment (meain) [reply src/internal/kopia/data_collection.go:43]: |||
|---|---|
|Small files within multiple directories|Lot of tiny files in a single directory|

Our memory profiles also seem to be cleaner with no mention of `produceRestoreCollection` anymore.

**Comment by meain** [comment]
@ashmrtn With the last 3 commits, I've changed it so that if we fail to read the metadata items we resort to a non-incremental backup. I did notice that we fail outright if we fail to parse the read data failing the backup. Should we change that behaviour as well?

File: src/internal/connector/data_collections.go
Comment (meain) [inline src/internal/connector/data_collections.go:83]: TODO: Can we safely assume there will be no previous backup if the service is not enabled?

File: src/internal/connector/onedrive/collections.go
Comment (meain) [inline src/internal/connector/onedrive/collections.go:234]: As of now, if we fail to fetch any metadata item, we revert to a non-incremental backup. We could maybe ignore failing to pull `delta`?

Comment (ashmrtn) [reply src/internal/connector/onedrive/collections.go:234]: let's go with this for now. It's alright to be a bit conservative and we can come back and loosen the restrictions later on

**Review by ashmrtn — APPROVED** [review]

**Pull request merged**
